rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,393 @@
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Case Study: Matcher
+
+The `matcher` plugin is an analyzer plugin that operator on the input stream. We
+can also rewrite the matcher use cases in term of pipelines, albeit with a
+slight change in CLI.
+
+#### Local execution
+
+Let's assume that the matcher introduces the following new pipeline operators:
+
+- `build<Arrow, Matcher>`: consumes a data stream to produce a matcher state object
+- `match<Arrow, Arrow>`: produces sightings for the configured matchers
+
+Now consider the use case of constructing a matcher from CSV data:
+
+```bash
+vast exec 'from file:///tmp/blacklist.csv |
+           read csv |
+           build --extract=net.src.ip' > matcher.flatbuf
+```
+
+After having constructed a matcher, we can now use it in a pipeline for
+matching:
+
+```bash
+vast exec 'from kafka -t /zeek/conn |
+           read zeek |
+           match --state matcher.flatbuf --on=:addr
+```
+
+The idea is that `match` gets the opaque matcher state as input. (We may
+generalize this builder pattern to deploy arbitrarily complex detections, e.g.,
+pre-trained models.)
+
+#### Remote execution
+
+In the above example, we passed in a state file via `--state`. This is a
+different pattern from today's matchers that run remotely. Assuming that we have
+a mechanism for remote state management (specifically matcher state), we can
+apply our existing UX pattern of remote matcher management.
+
+Let's assume the matcher plugin provides the `matcher` command for state
+management, just like today. To load the matcher state into a remote matcher, we
+nothing would change:
+
+```bash
+vast matcher load foo < matcher.flatbuf
+```
+
+Assuming this operations updates the state of matcher `foo`, we can then
+reference it in a remote pipeline:
+
+```bash
+vast spawn 'live | match --on=:addr foo'
+```
+
+Here, `live` is a dummy operator that represents the ingest path.
+
+### Pipeline Management
+
+In many cases we want to run pipelines remotely, independent of where the client
+is. This begs the challenge: how do we manage pipelines? As with all other VAST
+functions, there could be a client command.
+
+```bash
+# Show all pipelines:
+vast pipeline list
+```
+
+Create remote pipelines:
+
+```bash
+vast pipeline create foo 'where #type == "suricata.flow"'
+vast pipeline list
+# Output:
+# foo: where #type == "suricata.flow"
+```
+
+The idea would be that only clients can manipulate server state. This fixes
+the mess of having server-side YAML that goes out of sync with the server
+state when new things happen, e.g., new schemas arrive or new matchers get
+spawned.
+
+From now on, everything is client-side managed *only*. No more server-side
+state configuration beyond options/settings. This is where you can
+have your YAML for declarative ops:
```

Comment:
I think there's been a separate discussion about where things should be managed, but I think there's a use case for deploying a VAST node with predefined pipelines. Maybe those could be immutable, and if you need something different you just create another one with a different name?

No strong opinions here, though - it could also be something added later if we see demand.

Reply:
The way I envision the UX is that you spin up a VAST node, then run a client command to load a pre-defined set of pipelines (e.g., from a YAML file).

The optimization of this would be providing the configuration at startup time, but then we're back to the UX issue of "what happens when I want to make changes to the pipelines afterwards".

Reply:
Agree. I close this and refer back to "adding it later if we see demand".